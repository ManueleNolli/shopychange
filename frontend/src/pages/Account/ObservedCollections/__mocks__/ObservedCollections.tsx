import { divWithoutChildrenMock } from '../../../../../mocksUtils'

const ObservedCollectionsMock = () =>
  divWithoutChildrenMock('ObservedCollections')
export default ObservedCollectionsMock
