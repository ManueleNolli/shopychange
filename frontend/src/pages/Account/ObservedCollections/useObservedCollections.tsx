import { useEffect, useState } from 'react'
import {
  AddNFTCollectionsMutation,
  RemoveNFTCollectionsMutation,
  contractObservedQuery,
  handlerMutation,
  handlerQueryErrorCatch,
} from '../../../services/BackendService'
import useFetching from '../../../hooks/useFetching'
import { useUserContext } from '../../../context/userContext'

export default function useObservedCollections() {
  const { userAddress } = useUserContext()
  const {
    isFetching: isObservedFetching,
    setIsFetching: setIsObservedFetching,
    isError: isObservedError,
    setIsError: setIsObservedError,
  } = useFetching()
  const [observedCollections, setObservedCollections] = useState<string[]>([])

  const fetchObservedCollections = async () => {
    if (!userAddress) return

    const response = await handlerQueryErrorCatch(
      contractObservedQuery(userAddress)
    )

    if (response.error) {
      setIsObservedError(true)
      setIsObservedFetching(false)
      return
    }

    const observedCollections = response.result.contractObserved
    setObservedCollections(observedCollections)
    setIsObservedFetching(false)
  }

  useEffect(() => {
    fetchObservedCollections()
  }, [])

  const handleSubmit = async (values: { observedCollections: string[] }) => {
    if (!userAddress) return

    const addedCollections = values.observedCollections.filter(
      (collection) => !observedCollections.includes(collection)
    )

    if (addedCollections.length > 0) {
      const responseAdd = await handlerMutation(
        AddNFTCollectionsMutation(addedCollections, userAddress)
      )

      if (responseAdd.error) {
        setIsObservedError(true)
        return
      }
    }

    const removedCollections = observedCollections.filter(
      (collection) => !values.observedCollections.includes(collection)
    )

    if (removedCollections.length > 0) {
      const responseRemove = await handlerMutation(
        RemoveNFTCollectionsMutation(removedCollections, userAddress)
      )

      if (responseRemove.error) {
        setIsObservedError(true)
        return
      }
    }

    fetchObservedCollections()
  }

  return {
    observedCollections,
    isObservedFetching,
    isObservedError,
    handleSubmit,
  }
}
