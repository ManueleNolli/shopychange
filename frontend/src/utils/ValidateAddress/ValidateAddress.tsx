// validate blockchain address

export const validateAddress = (address: string) => {
  const regex = /^0x[a-fA-F0-9]{40}$/
  if (!regex.test(address)) {
    return 'Invalid address'
  }
}
