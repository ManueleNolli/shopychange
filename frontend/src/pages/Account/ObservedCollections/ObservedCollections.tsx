import React from 'react'

import {
  Center,
  Box,
  Flex,
  FormControl,
  FormErrorMessage,
  FormLabel,
  Input,
} from '@chakra-ui/react'
import useObservedCollections from './useObservedCollections'
import { Formik, FieldArray, Field, Form } from 'formik'
import { MdDelete } from 'react-icons/md'
import { TiPlus } from 'react-icons/ti'
import { validateAddress } from '../../../utils/ValidateAddress/ValidateAddress'
import LoadingButton from '../../../components/LoadingButton/LoadingButton'
import ShadowButton from '../../../components/ShadowButton/ShadowButton'
import { colors } from '../../../styles/Colors'
import QueryContainer from '../../../utils/QueryContainer/QueryContainer'

export default function ObservedCollections() {
  const {
    observedCollections,
    isObservedFetching,
    isObservedError,
    handleSubmit,
  } = useObservedCollections()

  return (
    <Box mt="2vh">
      <QueryContainer isLoading={isObservedFetching} isError={isObservedError}>
        <Center>
          <Formik
            enableReinitialize
            onSubmit={handleSubmit}
            initialValues={{
              observedCollections:
                observedCollections.length > 0 ? observedCollections : [''],
            }}
          >
            {({ values }) => (
              <Form>
                <FieldArray
                  validateOnChange={true}
                  name={'observedCollections'}
                >
                  {({
                    push,
                    remove,
                  }: /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
                  any) => (
                    <Box
                      mt="2"
                      maxW={{ sm: '600px', md: '1000px' }}
                      w={{ sm: '100vw', md: '60vw', lg: '50vw' }}
                    >
                      <FormLabel>Royalties</FormLabel>
                      {values.observedCollections.map(
                        (value: string, indexAttribute: number) => (
                          <Flex
                            key={`observedCollections[${indexAttribute}]`}
                            mt="2"
                          >
                            <Box w="88%">
                              <Field
                                key={`observedCollections[${indexAttribute}]`}
                                name={`observedCollections[${indexAttribute}]`}
                                validate={(value: string) =>
                                  validateAddress(value)
                                }
                              >
                                {/* eslint-disable-next-line
                                @typescript-eslint/no-explicit-any */}
                                {({ field, form }: any) => (
                                  <FormControl
                                    isInvalid={
                                      form.errors.observedCollections &&
                                      form.touched.observedCollections &&
                                      form.errors.observedCollections[
                                        indexAttribute
                                      ]
                                    }
                                  >
                                    <Input
                                      {...field}
                                      id={`observedCollections[${indexAttribute}]`}
                                      placeholder="Collection Address"
                                    />
                                    <FormErrorMessage>
                                      {form.errors.observedCollections &&
                                        form.touched.observedCollections &&
                                        form.errors.observedCollections[
                                          indexAttribute
                                        ]}
                                    </FormErrorMessage>
                                  </FormControl>
                                )}
                              </Field>
                            </Box>

                            <Center w="8%">
                              <Box maxW="36px">
                                <MdDelete
                                  aria-label="remove-address"
                                  onClick={() => remove(indexAttribute)}
                                  fill="#FF4500"
                                  size={'80%'}
                                />
                              </Box>
                            </Center>
                          </Flex>
                        )
                      )}
                      <Box mt="2" h="2.5vh" w="2.5vh">
                        <Box maxW="24px">
                          <TiPlus
                            aria-label="add-address"
                            fill="#228B22"
                            size={'100%'}
                            onClick={() => push('')}
                          />
                        </Box>
                      </Box>
                    </Box>
                  )}
                </FieldArray>
                <Center>
                  <LoadingButton
                    isLoading={false}
                    loadingButtonProps={{
                      size: 'lg',
                      w: '18em',
                      maxW: '18em',
                    }}
                  >
                    <ShadowButton
                      label={'Update Observed Collections'}
                      bgColor={`${colors.sellChakraColor}.400`}
                      textColor={'white'}
                      shadowColor={colors.sellShadowColor}
                      hoverColor={`${colors.sellChakraColor}.500`}
                      focusColor={`${colors.sellChakraColor}.500`}
                      type="submit"
                      buttonProps={{
                        size: 'lg',
                        w: '18em',
                        maxW: '18em',
                      }}
                    />
                  </LoadingButton>
                </Center>
              </Form>
            )}
          </Formik>
        </Center>
      </QueryContainer>
    </Box>
  )
}
