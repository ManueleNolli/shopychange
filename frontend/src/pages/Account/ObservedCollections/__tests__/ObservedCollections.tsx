import React from 'react'
import { render, fireEvent, act, screen } from '@testing-library/react'
import ObservedCollections from '../ObservedCollections'
import useObservedCollections from '../useObservedCollections'
import { validateAddress } from '../../../../utils/ValidateAddress/ValidateAddress'

jest.mock('../useObservedCollections', () => jest.fn())

jest.mock('../../../../utils/ValidateAddress/ValidateAddress', () => ({
  validateAddress: jest.fn(),
}))

describe('ObservedCollections', () => {
  test('render correctly', () => {
    ;(useObservedCollections as jest.Mock).mockReturnValue({
      observedCollections: [],
      isObservedFetching: false,
      isObservedError: false,
      handleSubmit: jest.fn(),
    })

    const { container } = render(<ObservedCollections />)

    expect(container).toMatchSnapshot()
  })

  test('validate address', async () => {
    ;(useObservedCollections as jest.Mock).mockReturnValue({
      observedCollections: [],
      isObservedFetching: false,
      isObservedError: false,
      handleSubmit: jest.fn(),
    })
    ;(validateAddress as jest.Mock).mockReturnValue(true)

    const { container } = render(<ObservedCollections />)
    const input = container.querySelector(
      'input[name="observedCollections[0]"]'
    )

    // act to trigger validateAddress

    await act(async () => {
      fireEvent.change(input as Element, {
        target: { value: '0x123' },
      })
    })

    expect(validateAddress).toHaveBeenCalled()
  })

  test('should add addresses', async () => {
    ;(useObservedCollections as jest.Mock).mockReturnValue({
      observedCollections: [],
      isObservedFetching: false,
      isObservedError: false,
      handleSubmit: jest.fn(),
    })
    ;(validateAddress as jest.Mock).mockReturnValue(true)

    render(<ObservedCollections />)
    const addresses = screen.getAllByPlaceholderText('Collection Address')

    expect(addresses.length).toBe(1)

    const button = screen.getByLabelText('add-address')

    // act to add address
    await act(async () => {
      fireEvent.click(button)
    })

    const addressesAfterAdd =
      screen.getAllByPlaceholderText('Collection Address')

    expect(addressesAfterAdd.length).toBe(2)
  })

  test('should remove addresses', async () => {
    ;(useObservedCollections as jest.Mock).mockReturnValue({
      observedCollections: [],
      isObservedFetching: false,
      isObservedError: false,
      handleSubmit: jest.fn(),
    })
    ;(validateAddress as jest.Mock).mockReturnValue(true)

    render(<ObservedCollections />)
    const addresses = screen.getAllByPlaceholderText('Collection Address')

    expect(addresses.length).toBe(1)

    const button = screen.getByLabelText('remove-address')

    // act to remove address
    await act(async () => {
      fireEvent.click(button)
    })

    const addressesAfterAdd =
      screen.queryAllByPlaceholderText('Collection Address')

    expect(addressesAfterAdd.length).toBe(0)
  })
})
