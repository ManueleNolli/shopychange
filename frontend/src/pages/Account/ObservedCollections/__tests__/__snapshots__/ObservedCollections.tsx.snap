// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObservedCollections render correctly 1`] = `
<div>
  <div
    class="css-1g9j6we"
  >
    <div
      class="css-gmuwbf"
    >
      <form
        action="#"
      >
        <div
          class="css-jzuumy"
        >
          <label
            class="chakra-form__label css-uvho5g"
          >
            Royalties
          </label>
          <div
            class="css-bj7zsy"
          >
            <div
              class="css-13phf6t"
            >
              <div
                class="chakra-form-control css-0"
                role="group"
              >
                <input
                  class="chakra-input css-0"
                  id="observedCollections[0]"
                  name="observedCollections[0]"
                  placeholder="Collection Address"
                  value=""
                />
              </div>
            </div>
            <div
              class="css-1dortbl"
            >
              <div
                class="css-cse5we"
              >
                <svg
                  aria-label="remove-address"
                  fill="#FF4500"
                  height="80%"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="80%"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                  <path
                    d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                  />
                </svg>
              </div>
            </div>
          </div>
          <div
            class="css-98x68k"
          >
            <div
              class="css-913hcs"
            >
              <svg
                aria-label="add-address"
                baseProfile="tiny"
                fill="#228B22"
                height="100%"
                stroke="currentColor"
                stroke-width="0"
                version="1.2"
                viewBox="0 0 24 24"
                width="100%"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M18 10h-4v-4c0-1.104-.896-2-2-2s-2 .896-2 2l.071 4h-4.071c-1.104 0-2 .896-2 2s.896 2 2 2l4.071-.071-.071 4.071c0 1.104.896 2 2 2s2-.896 2-2v-4.071l4 .071c1.104 0 2-.896 2-2s-.896-2-2-2z"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="css-gmuwbf"
        >
          <button
            class="chakra-button css-g5wf5l"
            data-testid="onClick-button"
            style="white-space: initial;"
            type="submit"
          >
            Update Observed Collections
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
